utils/fork-tree/src/lib.rs

```diff
@@ -755,48 +769,64 @@ mod node_implementation {
 				  F: Fn(&H, &H) -> Result<bool, E>,
 				  P: Fn(&V) -> bool,
 		{
-			// stop searching this branch
-			if *number < self.number {
-				return Ok(FindOutcome::Failure(false));
-			}
+			let mut stack: Vec<(&Self, Option<usize>, bool)> = vec![(self, None, true)];
 
-			let mut known_descendent_of = false;
-
-			// continue depth-first search through all children
-			for (i, node) in self.children.iter().enumerate() {
-				// found node, early exit
-				match node.find_node_index_where(hash, number, is_descendent_of, predicate)? {
-					FindOutcome::Abort => return Ok(FindOutcome::Abort),
-					FindOutcome::Found(mut x) => {
-						x.push(i);
-						return Ok(FindOutcome::Found(x))
-					},
-					FindOutcome::Failure(true) => {
-						// if the block was a descendent of this child,
-						// then it cannot be a descendent of any others,
-						// so we don't search them.
-						known_descendent_of = true;
-						break;
-					},
-					FindOutcome::Failure(false) => {},
+			let mut found = false;
+			let mut indices: Option<Vec<usize>> = None;
+
+			while let Some((node, index, first_time)) = stack.pop() {
+				// stop searching this branch
+				if *number < node.number {
+					continue;
 				}
-			}
 
-			// node not found in any of the descendents, if the node we're
-			// searching for is a descendent of this node then we will stop the
-			// search here, since there aren't any more children and we found
-			// the correct node so we don't want to backtrack.
-			let is_descendent_of = known_descendent_of || is_descendent_of(&self.hash, hash)?;
-			if is_descendent_of {
-				// if the predicate passes we return the node
-				if predicate(&self.data) {
-					return Ok(FindOutcome::Found(Vec::new()));
+				if first_time && !found && !node.children.is_empty() {
+					stack.push((node, index, false));
+
+					for (i, child) in node.children.iter().enumerate().rev() {
+						stack.push((child, Some(i), true));
+					}
+
+					continue;
+				}
+
+				let is_descendent_of = is_descendent_of(&node.hash, hash)?;
+
+				if is_descendent_of {
+					found |= predicate(&node.data);
+
+					if !found {
+						while stack.last().map(|(.., first)| *first).unwrap_or(false) {
+							stack.pop();
+						}
```

Comment:
So here we pop all nodes in their first cycle through the stack off it, which is the equivalent of the `break` in the recursive code. Strangely, if I change this to be `return Ok(FindOutcome::Failure(false))`, then `ashley-load-sync-state` does not panic, but continuously prints `2020-10-19 15:44:18  💔 Verification failed for block 0x216fd2aae1a4807eda03558a95d87712e3e2a6092964b2a880b70850d7dda077 received from peer: 12D3KooWLK2gMLhWsYJzjW3q35zAs9FDDVqfqVfVuskiGZGRSMvR, "Could not fetch epoch at 0x659d978076da389f0089d04c9851812fd68da5c9aa293efe5206555729764897"
`.